### PR TITLE
Shallow project samples method for REST

### LINKED ISSUE #214: Slow listing REST /project/x/samples for large projects
Listing a large project's samples can take a very long time in the REST API.  Large being around ~7k samples.  At face value listing this many samples shouldn't be a problem, but joins in the database appear to be what's taking the majority of the time.

Most likely candidate is the metadata for a sample: https://github.com/phac-nml/irida/blob/development/src/main/java/ca/corefacility/bioinformatics/irida/model/sample/Sample.java#L145.  Since this is `FetchType.EAGER` loaded it's doing a separate `SELECT` for every sample to load the metadata.

### PR DESCRIPTION
**Pull request by tom114**
Added a custom repository method to shallowly list the samples for a project.  The issue with #214 seemed to be the huge number of queries going on to list the metadata for the samples in a project.  For just getting the samples in the REST API we didn't need all that information.

Fixes #214

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryImpl.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	public List<Sample> getSamplesForProjectShallow(Project project) {
+		NamedParameterJdbcTemplate tmpl = new NamedParameterJdbcTemplate(dataSource);
+		MapSqlParameterSource parameters = new MapSqlParameterSource();
+
+		//query to read samples for a project
+		String queryString = "select s.id, s.createdDate, s.modifiedDate, s.description, s.sampleName, s.collectedBy, s.geographicLocationName, s.isolate, s.isolationSource, s.latitude, s.longitude, s.organism, s.strain, s.collectionDate, null as remote_status FROM sample s INNER JOIN project_sample p ON p.sample_id=s.id WHERE p.project_id=:project";
+
+		parameters.addValue("project", project.getId());
+
+		return tmpl.query(queryString, parameters, new BeanPropertyRowMapper(Sample.class));
```
Comment (apetkau) [inline src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryImpl.java]: It looks like the `createdDate` for a sample is not being set correctly in these results. It's instead being set to today's date.

## Results from API for current IRIDA release

```
```
{
  "resource" : {
    "resources" : [ {
      "createdDate" : 1544652180042,
      "modifiedDate" : 1538512517000,
      "sampleName" : "08-5578",
...
```

The `createdDate` corresponds to today's date **Wed, 12 Dec 2018 22:03:00 +0000**

## Results from API from this code

```
{
  "resource" : {
    "resources" : [ {
      "createdDate" : 1535122022000,
      "modifiedDate" : 1538512517000,
      "sampleName" : "08-5578",
```

The `createdDate` corresponds to the correct sample date **Fri, 24 Aug 2018 14:47:02 +0000**

Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryImpl.java]: Looks like the `BeanPropertyRowMapper` wasn't properly overriding the `createdDate` from the constructor.  Switched to using a manual `RowMapper` in ffc28e0.  Still just as fast.

Comment (apetkau) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryImpl.java]: Awesome, thanks :+1:

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryCustom.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import ca.corefacility.bioinformatics.irida.model.project.Project;
+import ca.corefacility.bioinformatics.irida.model.sample.Sample;
+
+import java.util.List;
+
+/**
+ * Custom methods for getting {@link Sample}s.  This can be used to create custom, higher speed methods for specific
+ * {@link Sample} related tasks.
+ */
+public interface SampleRepositoryCustom {
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryCustom.java]: I think I prefer leaving the `Custom` name for a couple reasons:
1. It's the naming scheme we've used for other "custom" repos in the project (see `AnalysisSubmissionRepositoryCustom.java`, `MetadataFieldRepositoryCustom.java`, `UserRepositoryCustom.java`.  I'm pretty sure this used to be a required naming scheme for Spring Data.
2. This wouldn't necessarily always just have "high performance" methods.  I could see this being used for metadata stuff in the future for queries that are too complex for JPA.

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java [hunk 3]
```diff
@@ -182,13 +180,11 @@ public ModelMap getProjectSamples(@PathVariable Long projectId) {
 
 		ModelMap modelMap = new ModelMap();
 		Project p = projectService.read(projectId);
-		List<Join<Project, Sample>> relationships = sampleService.getSamplesForProject(p);
-
-		ResourceCollection<Sample> sampleResources = new ResourceCollection<>(relationships.size());
+		List<Sample> samples = sampleService.getSamplesForProjectShallow(p);
```
[the remark this answers is not in the document]
Comment (tom114) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java]: Yeah wow was that ever out of date.  Update in 4ba1fc6b66

File: src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryCustom.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (apetkau) [reply src/main/java/ca/corefacility/bioinformatics/irida/repositories/sample/SampleRepositoryCustom.java]: Okay, that makes sense to me.

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (apetkau) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/projects/RESTProjectSamplesController.java]: Thanks.

**Comment by apetkau** [comment]
Everything looks great to me. Thanks.

**Pull request merged**